source/opt/dead_branch_elim_pass.cpp

```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
-    auto dbi = bi;
-    ++dbi;
-    uint32_t dLabId = (*dbi)->id();
-    while (dLabId != mergeLabId) {
-      if (!HasNonPhiNonBackedgeRef(dLabId)) {
-        // Kill use/def for all instructions and mark block for elimination
-        backedges_.erase((*dbi)->terminator());
-        KillAllInsts(*dbi);
-        elimBlocks.insert(*dbi);
+    // Don't simplify branches of continue blocks. A path from the continue to
+    // the merge is required.
+    // TODO(alan-baker): They can be simplified iff there remains a path to the
+    // backedge. Structured control flow should guarantee one path hits the
+    // backedge, but I've removed the requirement for structured control flow
+    // from this pass.
+    bool simplify = liveLabId != 0 && !continues.count(block);
+
+    if (simplify) {
+      modified = true;
+      // Replace with unconditional branch.
+      // Remove the merge instruction if it is a selection merge.
+      AddBranch(liveLabId, block);
+      context()->KillInst(terminator);
+      ir::Instruction* mergeInst = block->GetMergeInst();
+      if (mergeInst->opcode() == SpvOpSelectionMerge) {
+        context()->KillInst(mergeInst);
       }
-      ++dbi;
-      dLabId = (*dbi)->id();
+      stack.push_back(get_parent_block(liveLabId));
+    } else {
+      // All successors are live.
+      block->ForEachSuccessorLabel(
+          [&stack, get_parent_block](const uint32_t label) {
+            stack.push_back(get_parent_block(label));
+          });
     }
+  }
 
-    // If merge block is unreachable, continue eliminating blocks until
-    // a live block or last block is reached.
-    while (!HasNonPhiNonBackedgeRef(dLabId)) {
-      backedges_.erase((*dbi)->terminator());
-      KillAllInsts(*dbi);
-      elimBlocks.insert(*dbi);
-      ++dbi;
-      if (dbi == structuredOrder.end()) break;
-      dLabId = (*dbi)->id();
+  // Check for unreachable merge and continue blocks with live headers, those
+  // blocks must remain. Continues are tracked separately so that when updating
+  // live phi nodes with an edge from a continue I can replace it with an
+  // undef (because I still clobber the continue block).
+  std::unordered_set<ir::BasicBlock*> unreachableMerges;
+  std::unordered_map<ir::BasicBlock*, ir::BasicBlock*> unreachableContinues;
+  for (auto block : liveBlocks) {
+    uint32_t mergeId = block->MergeBlockIdIfAny();
+    if (mergeId != 0) {
+      ir::BasicBlock* mergeBlock = get_parent_block(mergeId);
+      if (!liveBlocks.count(mergeBlock)) {
+        unreachableMerges.insert(mergeBlock);
+      }
+      uint32_t contId = block->ContinueBlockIdIfAny();
+      if (contId != 0) {
+        ir::BasicBlock* contBlock = get_parent_block(contId);
+        if (!liveBlocks.count(contBlock)) {
+          unreachableContinues[contBlock] = block;
+        }
+      }
     }
+  }
 
-    // If last block reached, look for next dead branch
-    if (dbi == structuredOrder.end()) continue;
+  for (auto& block : *func) {
```

Comment:
Given the length of this function, you need a comment saying what this phase does.    I guess I suggest moving the comment starting with "Fix phis" out to the above the iteration over the blocks.

Then rehprase it as   "Fix phis in reachable blocks so that ...."

Reply:
I would suggest splitting it up into separate functions.  I found the function long as well.  So there would be a `FindLiveBlocks`, `MarkStructuredTargetsAsLive`, `UpdatePhis`, and `DeleteDeadBlocks`.  Or something like that.  it would make it much easier to read, at least for me.

Reply:
I'll do some refactoring.